File: src/year2021/month1/no94/BinaryTreeInorderTraversal.java
package year2021.month1.no94;

import java.util.*;

public class BinaryTreeInorderTraversal {
    public static void main(String[] args) {
        TreeNode root1 = new TreeNode(1, null, new TreeNode(2, new TreeNode(3), null));
        TreeNode root2 = null;
        TreeNode root3 = new TreeNode(1);
        TreeNode root4 = new TreeNode(1, new TreeNode(2), null);
        TreeNode root5 = new TreeNode(1, null, new TreeNode(2));
        System.out.println(inorderTraversal2(root1));
        System.out.println(inorderTraversal2(root2));
        System.out.println(inorderTraversal2(root3));
        System.out.println(inorderTraversal2(root4));
        System.out.println(inorderTraversal2(root5));
    }

    private static List<Integer> inorderTraversal2(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Deque<TreeNode> stack = new LinkedList<>();
//        用Set代表节点是否访问过
        Set<TreeNode> set = new HashSet<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode pop = stack.pop();
            if (pop == null) {
                continue;
            }
            if (set.contains(pop)) {
//                节点已访问过，则输出其值
                result.add(pop.val);
                continue;
            }
//            栈为：先进后出，而中序遍历为：左中右，故进栈顺序为：右中左，且需要将当前访问元素pop标记为已访问，并再次入栈
            stack.push(pop.right);
            stack.push(pop);
            set.add(pop);
            stack.push(pop.left);
        }
        return result;
    }

    private static List<Integer> inorderTraversal1(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode node = root;
        while (node != null || !stack.isEmpty()) {
            while (node != null) {
                stack.push(node);
                node = node.left;
            }
            node = stack.pop();
            result.add(node.val);
            node = node.right;
        }
        return result;
    }

    public static List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        dfs(root, result);
        return result;
    }

    private static void dfs(TreeNode root, List<Integer> result) {
        if (root == null) {
            return;
        }
        dfs(root.left, result);
        result.add(root.val);
        dfs(root.right, result);
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/*
* 给定一个二叉树的根节点 root ，返回它的 中序遍历。



示例 1：


输入：root = [1,null,2,3]
输出：[1,3,2]
示例 2：

输入：root = []
输出：[]
示例 3：

输入：root = [1]
输出：[1]
示例 4：


输入：root = [1,2]
输出：[2,1]
示例 5：


输入：root = [1,null,2]
输出：[1,2]


提示：

树中节点数目在范围 [0, 100] 内
-100 <= Node.val <= 100


进阶:递归算法很简单，你可以通过迭代算法完成吗？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/binary-tree-inorder-traversal
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month5/no94/BinaryTreeInorderTraversal.go
package main

import "fmt"

func main() {
	root1 := &TreeNode{
		Val: 1,
		Right: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 3,
			},
		},
	}
	root2 := (*TreeNode)(nil)
	root3 := &TreeNode{
		Val: 1,
	}
	println(fmt.Sprintf("%+v", inorderTraversal(root1)))
	println(fmt.Sprintf("%+v", inorderTraversal(root2)))
	println(fmt.Sprintf("%+v", inorderTraversal(root3)))
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Stack, times: O(n), space: O(h)
func inorderTraversal(root *TreeNode) (ans []int) {
	stack := make([]*TreeNode, 0)
	p := root
	for len(stack) > 0 || p != nil {
		for p != nil {
			stack = append(stack, p)
			p = p.Left
		}
		pop := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		ans = append(ans, pop.Val)
		p = pop.Right
	}
	return ans
}

// dfs, times: O(n), space: O(h)
func inorderTraversal1(root *TreeNode) []int {
	return dfs(root, make([]int, 0))
}

func dfs(root *TreeNode, arr []int) []int {
	if root == nil {
		return arr
	}
	arr = dfs(root.Left, arr)
	arr = append(arr, root.Val)
	arr = dfs(root.Right, arr)
	return arr
}

/*
给定一个二叉树的根节点 root ，返回 它的 中序 遍历 。

示例 1：
输入：root = [1,null,2,3]
输出：[1,3,2]

示例 2：
输入：root = []
输出：[]

示例 3：
输入：root = [1]
输出：[1]

提示：
树中节点数目在范围 [0, 100] 内
-100 <= Node.val <= 100
*/
